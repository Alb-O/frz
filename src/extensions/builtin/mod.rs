pub mod attributes;
pub mod files;
pub mod logger;

use crate::extensions::api::{ExtensionCatalog, ExtensionCatalogError, ExtensionDescriptor};

pub fn register_builtin_extensions(
    catalog: &mut ExtensionCatalog,
) -> Result<(), ExtensionCatalogError> {
    catalog.register_package(attributes::bundle())?;
    catalog.register_package(files::bundle())?;
    catalog.register_package(logger::bundle())?;
    Ok(())
}

pub fn descriptors() -> &'static [&'static ExtensionDescriptor] {
    &BUILTIN_DESCRIPTORS
}

static BUILTIN_DESCRIPTORS: [&ExtensionDescriptor; 3] = [
    &attributes::ATTRIBUTE_DESCRIPTOR,
    &files::FILE_DESCRIPTOR,
    &logger::LOGGER_DESCRIPTOR,
];
