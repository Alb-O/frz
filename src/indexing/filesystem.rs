use std::collections::{BTreeMap, HashSet};
use std::path::PathBuf;
use std::sync::Arc;
use std::sync::mpsc::{self, Receiver, Sender};
use std::thread;
use std::time::{Duration, Instant};

use anyhow::Result;
use ignore::{DirEntry, Error as IgnoreError, WalkBuilder, WalkState};

use crate::types::tags_for_relative_path;
use crate::types::{FacetRow, FileRow, SearchData};

use super::{IndexUpdate, ProgressSnapshot};

#[path = "filesystem_cache.rs"]
mod cache;
use cache::CacheHandle;
use cache::CacheWriter;

const MIN_BATCH_SIZE: usize = 32;
const MAX_BATCH_SIZE: usize = 1_024;
const DISPATCH_INTERVAL: Duration = Duration::from_millis(120);

#[derive(Debug, Clone)]
pub struct FilesystemOptions {
    pub include_hidden: bool,
    pub follow_symlinks: bool,
    pub respect_ignore_files: bool,
    pub git_ignore: bool,
    pub git_global: bool,
    pub git_exclude: bool,
    pub global_ignores: Vec<String>,
    pub threads: Option<usize>,
    pub max_depth: Option<usize>,
    pub allowed_extensions: Option<Vec<String>>,
    pub context_label: Option<String>,
}

impl Default for FilesystemOptions {
    fn default() -> Self {
        Self {
            include_hidden: true,
            follow_symlinks: false,
            respect_ignore_files: true,
            git_ignore: true,
            git_global: true,
            git_exclude: true,
            global_ignores: vec![
                ".git".to_string(),
                "node_modules".to_string(),
                "target".to_string(),
                ".venv".to_string(),
                ".cache".to_string(),
            ],
            threads: None,
            max_depth: None,
            allowed_extensions: None,
            context_label: None,
        }
    }
}

pub(crate) fn spawn_filesystem_index(
    root: PathBuf,
    mut options: FilesystemOptions,
) -> Result<(SearchData, Receiver<IndexUpdate>)> {
    let (tx, rx) = mpsc::channel();

    let cache_handle = CacheHandle::resolve(&root, &options);
    let mut data = SearchData::new();

    if options.context_label.is_none() {
        options.context_label = Some(root.display().to_string());
    }
    data.context_label = options.context_label.clone();

    let should_reset = cache_handle.is_some();
    let cache_handle_for_thread = cache_handle.clone();
    let context_label = options.context_label.clone();

    thread::spawn(move || {
        let mut reindex_delay = Duration::ZERO;

        if let Some(handle) = cache_handle_for_thread.as_ref() {
            if let Some(entry) = handle.load() {
                reindex_delay = entry.reindex_delay();

                let mut data = entry.data;
                if data.context_label.is_none() {
                    data.context_label = context_label.clone();
                }

                let files: Arc<[FileRow]> = data.files.clone().into();
                let facets: Arc<[FacetRow]> = data.facets.clone().into();
                let progress = ProgressSnapshot {
                    indexed_facets: facets.len(),
                    indexed_files: files.len(),
                    total_facets: Some(facets.len()),
                    total_files: Some(files.len()),
                    complete: false,
                };

                if !files.is_empty() || !facets.is_empty() {
                    let _ = tx.send(IndexUpdate {
                        files,
                        facets,
                        progress,
                        reset: true,
                        cached_data: Some(data),
                    });
                }
            }
        }

        if !reindex_delay.is_zero() {
            thread::sleep(reindex_delay);
        }

        let (file_tx, file_rx) = mpsc::channel::<FileRow>();
        let walker_root = Arc::new(root);
        let threads = options
            .threads
            .filter(|threads| *threads > 0)
            .unwrap_or_else(|| {
                std::thread::available_parallelism().map_or(1, std::num::NonZeroUsize::get)
            });
        let extension_filter = options.allowed_extensions.as_ref().map(|extensions| {
            Arc::new(
                extensions
                    .iter()
                    .map(|ext| normalize_extension(ext))
                    .filter(|ext| !ext.is_empty())
                    .collect::<HashSet<_>>(),
            )
        });
        let update_tx = tx;

        let cache_writer = cache_handle_for_thread
            .as_ref()
            .and_then(|handle| handle.writer(context_label.clone()));
        let aggregator = thread::spawn(move || {
            let mut batcher = UpdateBatcher::new(should_reset, cache_writer);

            while let Ok(file) = file_rx.recv() {
                batcher.record_file(file);

                if batcher.should_flush() && batcher.flush(&update_tx, false).is_err() {
                    return None;
                }
            }

            batcher.finalize(&update_tx).unwrap_or_default()
        });

        let global_ignores = options.global_ignores.clone();

        WalkBuilder::new(walker_root.as_path())
            .hidden(!options.include_hidden)
            .follow_links(options.follow_symlinks)
            .git_ignore(options.git_ignore)
            .git_global(options.git_global)
            .git_exclude(options.git_exclude)
            .ignore(options.respect_ignore_files)
            .parents(true)
            .threads(threads)
            .max_depth(options.max_depth)
            .build_parallel()
            .run(|| {
                let sender = file_tx.clone();
                let root = Arc::clone(&walker_root);
                let extension_filter = extension_filter.clone();
                let global_ignores = global_ignores.clone();
                Box::new(move |entry: Result<DirEntry, IgnoreError>| {
                    if let Ok(entry) = entry {
                        let Some(file_type) = entry.file_type() else {
                            return WalkState::Continue;
                        };
                        if !file_type.is_file() {
                            return WalkState::Continue;
                        }

                        let path = entry.path();
                        // Skip global ignore paths
                        if path.components().any(|comp| {
                            comp.as_os_str()
                                .to_str()
                                .map(|s| global_ignores.iter().any(|g| g == s))
                                .unwrap_or(false)
                        }) {
                            return WalkState::Continue;
                        }
                        let relative = path.strip_prefix(root.as_path()).unwrap_or(path);
                        if let Some(filter) = extension_filter.as_ref() {
                            let extension = relative
                                .extension()
                                .and_then(|ext| ext.to_str())
                                .map(|ext| ext.to_ascii_lowercase());
                            if extension.as_ref().is_none_or(|ext| !filter.contains(ext)) {
                                return WalkState::Continue;
                            }
                        }
                        let tags = tags_for_relative_path(relative);
                        let relative_display = relative.to_string_lossy().replace('\\', "/");
                        let file = FileRow::new(relative_display, tags);
                        if sender.send(file).is_err() {
                            return WalkState::Quit;
                        }
                    }

                    WalkState::Continue
                })
            });

        drop(file_tx);
        if let Ok(Some(writer)) = aggregator.join() {
            let _ = writer.finish();
        }
    });

    Ok((data, rx))
}

struct UpdateBatcher {
    facet_counts: BTreeMap<String, usize>,
    pending_facets: BTreeMap<String, usize>,
    pending_files: Vec<FileRow>,
    indexed_files: usize,
    last_dispatch: Instant,
    emit_reset: bool,
    cache_writer: Option<CacheWriter>,
}

impl UpdateBatcher {
    fn new(emit_reset: bool, cache_writer: Option<CacheWriter>) -> Self {
        Self {
            facet_counts: BTreeMap::new(),
            pending_facets: BTreeMap::new(),
            pending_files: Vec::new(),
            indexed_files: 0,
            last_dispatch: Instant::now(),
            emit_reset,
            cache_writer,
        }
    }

    fn record_file(&mut self, file: FileRow) {
        if let Some(writer) = &mut self.cache_writer {
            writer.record(&file);
        }

        for tag in &file.tags {
            let count = self.facet_counts.entry(tag.clone()).or_insert(0);
            *count += 1;
            self.pending_facets.insert(tag.clone(), *count);
        }

        self.indexed_files += 1;
        self.pending_files.push(file);
    }

    fn should_flush(&self) -> bool {
        if self.pending_files.len() >= batch_size_for(self.indexed_files) {
            return true;
        }

        if !self.emit_reset && self.pending_files.is_empty() && self.pending_facets.is_empty() {
            return false;
        }

        self.last_dispatch.elapsed() >= DISPATCH_INTERVAL
    }

    fn flush(
        &mut self,
        tx: &Sender<IndexUpdate>,
        complete: bool,
    ) -> Result<(), mpsc::SendError<IndexUpdate>> {
        if !complete
            && !self.emit_reset
            && self.pending_files.is_empty()
            && self.pending_facets.is_empty()
        {
            return Ok(());
        }

        let files_vec = std::mem::take(&mut self.pending_files);
        let files: Arc<[FileRow]> = files_vec.into();
        let facets: Arc<[FacetRow]> = if self.pending_facets.is_empty() {
            Arc::default()
        } else {
            let collected: Vec<FacetRow> = self
                .pending_facets
                .iter()
                .map(|(name, count)| FacetRow::new(name.clone(), *count))
                .collect();
            self.pending_facets.clear();
            collected.into()
        };

        let progress = ProgressSnapshot {
            indexed_facets: self.facet_counts.len(),
            indexed_files: self.indexed_files,
            total_facets: complete.then_some(self.facet_counts.len()),
            total_files: complete.then_some(self.indexed_files),
            complete,
        };

        let reset = self.emit_reset;
        if self.emit_reset {
            self.emit_reset = false;
        }

        tx.send(IndexUpdate {
            files,
            facets,
            progress,
            reset,
            cached_data: None,
        })?;

        self.last_dispatch = Instant::now();
        Ok(())
    }

    fn finalize(
        self,
        tx: &Sender<IndexUpdate>,
    ) -> Result<Option<CacheWriter>, mpsc::SendError<IndexUpdate>> {
        let mut this = self;
        this.flush(tx, true)?;
        Ok(this.cache_writer)
    }
}

fn batch_size_for(indexed_files: usize) -> usize {
    if indexed_files < 1_024 {
        MIN_BATCH_SIZE
    } else if indexed_files < 16_384 {
        256
    } else {
        MAX_BATCH_SIZE
    }
}

fn normalize_extension(ext: &str) -> String {
    ext.trim().trim_start_matches('.').to_ascii_lowercase()
}
