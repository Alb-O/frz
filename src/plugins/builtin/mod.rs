mod facets;
mod files;

use super::SearchPluginRegistry;

pub(crate) fn register_builtin_plugins(registry: &mut SearchPluginRegistry) {
    registry.register(facets::FacetSearchPlugin);
    registry.register(files::FileSearchPlugin);
}
