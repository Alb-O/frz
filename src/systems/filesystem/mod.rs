#[cfg(feature = "fs")]
mod filesystem;

use std::sync::Arc;

use crate::types::{FacetRow, FileRow, SearchData};

#[cfg(not(feature = "fs"))]
use std::path::PathBuf;
#[cfg(not(feature = "fs"))]
use std::sync::mpsc::Receiver;

#[cfg(not(feature = "fs"))]
use anyhow::Result;
#[cfg(not(feature = "fs"))]
use anyhow::bail;

#[cfg(feature = "fs")]
pub use filesystem::FilesystemOptions;
#[cfg(feature = "fs")]
pub use filesystem::spawn_filesystem_index;

#[cfg(feature = "fs")]
pub mod plugin;

/// Updates emitted by the filesystem indexer as it discovers new entries.
#[cfg_attr(not(feature = "fs"), allow(dead_code))]
#[derive(Debug, Clone)]
pub struct IndexUpdate {
    pub files: Arc<[FileRow]>,
    pub facets: Arc<[FacetRow]>,
    pub progress: ProgressSnapshot,
    pub reset: bool,
    pub cached_data: Option<SearchData>,
}

/// Snapshot of the indexing progress suitable for updating the UI tracker.
#[cfg_attr(not(feature = "fs"), allow(dead_code))]
#[derive(Debug, Clone, Copy)]
pub struct ProgressSnapshot {
    pub indexed_facets: usize,
    pub indexed_files: usize,
    pub total_facets: Option<usize>,
    pub total_files: Option<usize>,
    pub complete: bool,
}

pub fn merge_update(data: &mut SearchData, update: &IndexUpdate) {
    if update.reset {
        data.files.clear();
        data.facets.clear();
    }

    if !update.files.is_empty() {
        data.files.extend(update.files.iter().cloned());
    }

    if !update.facets.is_empty() {
        for facet in update.facets.iter() {
            match data
                .facets
                .binary_search_by(|existing| existing.name.cmp(&facet.name))
            {
                Ok(index) => data.facets[index].count = facet.count,
                Err(index) => data.facets.insert(index, facet.clone()),
            }
        }
    }
}

#[cfg(not(feature = "fs"))]
pub fn spawn_filesystem_index(_root: PathBuf) -> Result<(SearchData, Receiver<IndexUpdate>)> {
    bail!("filesystem support is disabled; enable the `fs` feature");
}
