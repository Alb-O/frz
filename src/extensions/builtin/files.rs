use ratatui::layout::{Constraint, Layout, Rect};

use crate::extensions::api::{
    Contribution, ExtensionModule, ExtensionPackage, ExtensionQueryContext,
    ExtensionSelectionContext, Icon, IconProvider, IconResource, IconStore, SearchData, SearchMode,
    SearchSelection, SearchStream,
    contributions::{PreviewResource, SelectionResolver},
    descriptors::{
        ExtensionDataset, ExtensionDescriptor, ExtensionUiDefinition, TableContext, TableDescriptor,
    },
    stream_files,
};
use crate::previewers::FilePreviewer;
use crate::tui::tables::rows::build_file_rows;

pub const DATASET_KEY: &str = "files";

pub fn mode() -> SearchMode {
    SearchMode::from_descriptor(descriptor())
}

pub fn descriptor() -> &'static ExtensionDescriptor {
    &FILE_DESCRIPTOR
}

static FILE_DATASET: FileDataset = FileDataset;

pub static FILE_DESCRIPTOR: ExtensionDescriptor = ExtensionDescriptor {
    id: DATASET_KEY,
    ui: ExtensionUiDefinition {
        tab_label: "Files",
        mode_title: "File search",
        hint: "Type to filter files.",
        table_title: "Matching files",
        count_label: "Files",
    },
    dataset: &FILE_DATASET,
};

struct FileDataset;

impl FileDataset {
    fn default_headers() -> Vec<String> {
        vec!["Path".into(), "Tags".into(), "Score".into()]
    }

    fn default_widths() -> Vec<Constraint> {
        vec![
            Constraint::Percentage(60),
            Constraint::Percentage(30),
            Constraint::Length(8),
        ]
    }

    fn resolve_column_widths(
        area: Rect,
        widths: &[Constraint],
        selection_width: u16,
        column_spacing: u16,
    ) -> Vec<u16> {
        if widths.is_empty() {
            return Vec::new();
        }

        let layout_area = Rect {
            x: 0,
            y: 0,
            width: area.width,
            height: 1,
        };
        let [_, columns_area] =
            Layout::horizontal([Constraint::Length(selection_width), Constraint::Fill(0)])
                .areas(layout_area);

        Layout::horizontal(widths.to_vec())
            .spacing(column_spacing)
            .split(columns_area)
            .iter()
            .map(|rect| rect.width)
            .collect()
    }
}

impl ExtensionDataset for FileDataset {
    fn key(&self) -> &'static str {
        DATASET_KEY
    }

    fn total_count(&self, data: &SearchData) -> usize {
        data.files.len()
    }

    fn build_table<'a>(&self, context: TableContext<'a>) -> TableDescriptor<'a> {
        let widths = context.widths.cloned().unwrap_or_else(Self::default_widths);
        let column_widths = Self::resolve_column_widths(
            context.area,
            &widths,
            context.selection_width,
            context.column_spacing,
        );
        let headers = context
            .headers
            .cloned()
            .unwrap_or_else(Self::default_headers);
        let icon_provider = context.scope.resolve::<IconStore>();
        let rows = build_file_rows(
            context.filtered,
            context.scores,
            &context.data.files,
            context.highlight,
            context.highlight_style,
            Some(&column_widths),
            icon_provider,
        );
        TableDescriptor::new(headers, widths, rows)
    }
}

pub struct FileModule;

impl ExtensionModule for FileModule {
    fn descriptor(&self) -> &'static ExtensionDescriptor {
        descriptor()
    }

    fn stream(
        &self,
        query: &str,
        stream: SearchStream<'_>,
        context: ExtensionQueryContext<'_>,
    ) -> bool {
        stream_files(context.data(), query, stream, context.latest_query_id())
    }

    fn selection(
        &self,
        context: ExtensionSelectionContext<'_>,
        index: usize,
    ) -> Option<SearchSelection> {
        context
            .data()
            .files
            .get(index)
            .cloned()
            .map(SearchSelection::File)
    }
}

pub struct FilePackage {
    contributions: [Contribution; 4],
}

impl FilePackage {
    fn new_contributions() -> [Contribution; 4] {
        [
            Contribution::search_tab(descriptor(), FileModule),
            Contribution::preview_split(descriptor(), FilePreviewer::default()),
            Contribution::icons(descriptor(), FileIcons),
            Contribution::selection_resolver(descriptor(), FileSelectionResolver),
        ]
    }
}

impl Default for FilePackage {
    fn default() -> Self {
        Self {
            contributions: Self::new_contributions(),
        }
    }
}

impl ExtensionPackage for FilePackage {
    type Contributions<'a> = std::array::IntoIter<Contribution, 4>;

    fn contributions(&self) -> Self::Contributions<'_> {
        self.contributions.clone().into_iter()
    }
}

#[must_use]
pub fn bundle() -> FilePackage {
    FilePackage::default()
}

// Use the Font Awesome file glyph for entries without a devicon match.
const GENERIC_FILE_ICON: char = '\u{f016}';

#[derive(Clone, Copy)]
struct FileIcons;

impl IconProvider for FileIcons {
    fn icon_for(&self, resource: IconResource<'_>) -> Option<Icon> {
        match resource {
            IconResource::File(row) => {
                let icon = devicons::FileIcon::from(row.path.as_str());
                let glyph = if icon.icon == '*' {
                    GENERIC_FILE_ICON
                } else {
                    icon.icon
                };
                Some(Icon::from_hex(glyph, icon.color))
            }
        }
    }
}

#[derive(Clone, Copy)]
struct FileSelectionResolver;

impl SelectionResolver for FileSelectionResolver {
    fn resolve<'a>(
        &self,
        data: &'a SearchData,
        filtered: &'a [usize],
        selected: Option<usize>,
    ) -> Option<PreviewResource<'a>> {
        let index = selected?;
        let row_index = filtered.get(index).copied()?;
        data.files.get(row_index).map(PreviewResource::File)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::extensions::api::search::FileRow;

    #[test]
    fn dataset_key() {
        assert_eq!(FILE_DATASET.key(), DATASET_KEY);
    }

    #[test]
    fn unknown_extensions_use_generic_file_icon() {
        let provider = FileIcons;
        let row = FileRow::filesystem("file.zzz", Vec::<String>::new());

        let icon = provider
            .icon_for(IconResource::File(&row))
            .expect("file icons should exist");

        assert_eq!(icon, Icon::from_hex(GENERIC_FILE_ICON, "#7e8ea8"));
    }
}
