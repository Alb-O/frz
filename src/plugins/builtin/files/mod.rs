use crate::plugins::{
    PluginQueryContext, PluginSelectionContext, SearchPlugin,
    descriptors::{
        SearchPluginDataset, SearchPluginDescriptor, SearchPluginUiDefinition, TableContext,
        TableDescriptor,
    },
    systems::search::{SearchStream, stream_files},
};
use crate::types::{SearchData, SearchMode, SearchSelection};
use ratatui::layout::{Constraint, Layout, Rect};

use crate::ui::components::tables::rows::build_file_rows;

const DATASET_KEY: &str = "files";

pub fn mode() -> SearchMode {
    SearchMode::from_descriptor(descriptor())
}

pub fn descriptor() -> &'static SearchPluginDescriptor {
    &FILE_DESCRIPTOR
}

static FILE_DATASET: FileDataset = FileDataset;

pub(crate) static FILE_DESCRIPTOR: SearchPluginDescriptor = SearchPluginDescriptor {
    id: DATASET_KEY,
    ui: SearchPluginUiDefinition {
        tab_label: "Files",
        mode_title: "File search",
        hint: "Type to filter files. Press Tab to view facets.",
        table_title: "Matching files",
        count_label: "Files",
    },
    dataset: &FILE_DATASET,
};

struct FileDataset;

impl FileDataset {
    fn default_headers() -> Vec<String> {
        vec!["Path".into(), "Tags".into(), "Score".into()]
    }

    fn default_widths() -> Vec<Constraint> {
        vec![
            Constraint::Percentage(60),
            Constraint::Percentage(30),
            Constraint::Length(8),
        ]
    }

    fn resolve_column_widths(
        area: Rect,
        widths: &[Constraint],
        selection_width: u16,
        column_spacing: u16,
    ) -> Vec<u16> {
        if widths.is_empty() {
            return Vec::new();
        }

        let layout_area = Rect {
            x: 0,
            y: 0,
            width: area.width,
            height: 1,
        };
        let [_, columns_area] =
            Layout::horizontal([Constraint::Length(selection_width), Constraint::Fill(0)])
                .areas(layout_area);

        Layout::horizontal(widths.to_vec())
            .spacing(column_spacing)
            .split(columns_area)
            .iter()
            .map(|rect| rect.width)
            .collect()
    }
}

impl SearchPluginDataset for FileDataset {
    fn key(&self) -> &'static str {
        DATASET_KEY
    }

    fn total_count(&self, data: &SearchData) -> usize {
        data.files.len()
    }

    fn build_table<'a>(&self, context: TableContext<'a>) -> TableDescriptor<'a> {
        let widths = context.widths.cloned().unwrap_or_else(Self::default_widths);
        let column_widths = Self::resolve_column_widths(
            context.area,
            &widths,
            context.selection_width,
            context.column_spacing,
        );
        let headers = context
            .headers
            .cloned()
            .unwrap_or_else(Self::default_headers);
        let rows = build_file_rows(
            context.filtered,
            context.scores,
            &context.data.files,
            context.highlight,
            Some(&column_widths),
        );
        TableDescriptor::new(headers, widths, rows)
    }
}

pub(crate) struct FileSearchPlugin;

impl SearchPlugin for FileSearchPlugin {
    fn descriptor(&self) -> &'static SearchPluginDescriptor {
        descriptor()
    }

    fn stream(
        &self,
        query: &str,
        stream: SearchStream<'_>,
        context: PluginQueryContext<'_>,
    ) -> bool {
        stream_files(context.data(), query, stream, context.latest_query_id())
    }

    fn selection(
        &self,
        context: PluginSelectionContext<'_>,
        index: usize,
    ) -> Option<SearchSelection> {
        context
            .data()
            .files
            .get(index)
            .cloned()
            .map(SearchSelection::File)
    }
}
