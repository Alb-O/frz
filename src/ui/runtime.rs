use std::collections::VecDeque;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, mpsc};
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{Result, anyhow};
use ratatui::crossterm::event::{self, Event, KeyEventKind};

use frz_plugin_api::{SearchData, SearchOutcome};

use super::App;

/// Construct an [`App`] for the provided data and run it to completion.
pub fn run(data: SearchData) -> Result<SearchOutcome> {
    let mut app = App::new(data);
    app.run()
}

impl<'a> App<'a> {
    /// Pump the terminal event loop until the user exits with a result.
    pub fn run(&mut self) -> Result<SearchOutcome> {
        let mut terminal = ratatui::init();
        terminal.clear()?;

        self.hydrate_initial_results();

        let (event_tx, event_rx) = mpsc::channel();
        let event_loop_running = Arc::new(AtomicBool::new(true));
        let event_loop_flag = Arc::clone(&event_loop_running);

        let event_thread = thread::spawn(move || -> Result<()> {
            while event_loop_flag.load(Ordering::Relaxed) {
                if event::poll(Duration::from_millis(50))? {
                    let event = event::read()?;
                    if event_tx.send(event).is_err() {
                        break;
                    }
                }
            }
            Ok(())
        });

        let mut pending_events = VecDeque::new();

        let result: Result<SearchOutcome> = 'event_loop: loop {
            self.pump_index_updates();
            self.pump_search_results();
            self.throbber_state.calc_next();

            loop {
                match event_rx.try_recv() {
                    Ok(Event::Resize(_, _)) => {}
                    Ok(event) => pending_events.push_back(event),
                    Err(mpsc::TryRecvError::Empty) => break,
                    Err(mpsc::TryRecvError::Disconnected) => {
                        break 'event_loop Err(anyhow!("input event channel disconnected"));
                    }
                }
            }

            terminal.draw(|frame| self.draw(frame))?;

            let mut maybe_outcome = None;
            while let Some(event) = pending_events.pop_front() {
                match event {
                    Event::Key(key) if key.kind == KeyEventKind::Press => {
                        if let Some(outcome) = self.handle_key(key)? {
                            maybe_outcome = Some(outcome);
                            break;
                        }
                    }
                    Event::Resize(_, _) => {}
                    _ => {}
                }
            }

            if let Some(outcome) = maybe_outcome {
                break Ok(outcome);
            }

            thread::sleep(Duration::from_millis(16));
        };

        ratatui::restore();

        event_loop_running.store(false, Ordering::Relaxed);
        match event_thread.join() {
            Ok(join_result) => join_result?,
            Err(err) => std::panic::resume_unwind(err),
        }

        result
    }

    fn hydrate_initial_results(&mut self) {
        if !self.search.has_issued_query() {
            self.mark_query_dirty();
            self.request_search();
        }

        if let Some(timeout) = self.initial_results_timeout {
            let deadline = Instant::now() + timeout;
            self.initial_results_deadline = Some(deadline);
            while Instant::now() < deadline {
                self.pump_index_updates();
                self.pump_search_results();
                if !self.search.is_in_flight() {
                    break;
                }
                thread::sleep(Duration::from_millis(10));
            }
            self.pump_search_results();
        } else {
            self.initial_results_deadline = None;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::plugins::builtin::files;
    use crate::systems::filesystem::{IndexUpdate, ProgressSnapshot};
    use frz_plugin_api::search::Fs;
    use frz_plugin_api::{AttributeRow, FileRow, SearchData};
    use ratatui::{Terminal, backend::TestBackend};
    use std::io;
    use std::path::{Path, PathBuf};
    use std::sync::mpsc;

    fn sample_index_update() -> IndexUpdate {
        IndexUpdate {
            files: vec![
                FileRow::filesystem("src/lib.rs", ["src", "*.rs"]),
                FileRow::filesystem("src/main.rs", ["src", "*.rs"]),
                FileRow::filesystem("README.md", ["*.md"]),
            ]
            .into(),
            attributes: vec![
                AttributeRow::new("*.md", 1),
                AttributeRow::new("*.rs", 2),
                AttributeRow::new("src", 2),
            ]
            .into(),
            progress: ProgressSnapshot {
                indexed_attributes: 3,
                indexed_files: 3,
                total_attributes: Some(3),
                total_files: Some(3),
                complete: true,
            },
            reset: true,
            cached_data: None,
        }
    }

    #[test]
    fn initial_files_tab_render_captures_missing_results() {
        let mut app = App::new(SearchData::new());
        app.set_mode(files::mode());
        app.hydrate_initial_results();

        let (tx, rx) = mpsc::channel();
        app.set_index_updates(rx);
        tx.send(sample_index_update()).unwrap();

        app.pump_index_updates();
        app.pump_search_results();

        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal.draw(|frame| app.draw(frame)).unwrap();

        let view = {
            let backend = terminal.backend();
            backend.to_string()
        };
        insta::assert_snapshot!("initial_files_tab_render_captures_missing_results", view);

        assert!(
            app.filtered_len() > 0,
            "expected initial search results to populate without any user input"
        );
    }

    struct MassiveSyntheticFs {
        total: usize,
    }

    impl MassiveSyntheticFs {
        fn new(total: usize) -> Self {
            Self { total }
        }
    }

    struct MassiveIter {
        remaining: usize,
        index: usize,
    }

    impl Iterator for MassiveIter {
        type Item = io::Result<PathBuf>;

        fn next(&mut self) -> Option<Self::Item> {
            if self.index >= self.remaining {
                return None;
            }

            let dir = self.index / 512;
            let path = PathBuf::from(format!("dir_{dir:04}/file_{:06}.txt", self.index));
            self.index += 1;
            Some(Ok(path))
        }
    }

    impl Fs for MassiveSyntheticFs {
        type Iter = MassiveIter;

        fn walk(&self, _root: &Path) -> io::Result<Self::Iter> {
            Ok(MassiveIter {
                remaining: self.total,
                index: 0,
            })
        }
    }

    #[test]
    fn massive_filesystem_initial_load_shows_preview_snapshot() {
        const TOTAL_FILES: usize = 125_000;

        let fs = MassiveSyntheticFs::new(TOTAL_FILES);
        let data = SearchData::from_filesystem_with(&fs, Path::new("/synthetic")).unwrap();
        assert!(
            data.files.len() >= 100_000,
            "expected synthetic filesystem to exceed 100k entries"
        );
        let total_files = data.files.len();
        let total_attributes = data.attributes.len();
        const PREVIEW_SLICE: usize = 512;
        let preview_files: Vec<FileRow> = data.files.iter().take(PREVIEW_SLICE).cloned().collect();
        let preview_attributes = data.attributes.clone();
        drop(data);

        let mut app = App::new(SearchData::new());
        app.set_mode(files::mode());
        app.hydrate_initial_results();

        assert_eq!(
            app.filtered_len(),
            0,
            "no results should be visible before indexing begins"
        );

        let (tx, rx) = mpsc::channel();
        app.set_index_updates(rx);
        tx.send(IndexUpdate {
            files: preview_files.into(),
            attributes: preview_attributes.into(),
            progress: ProgressSnapshot {
                indexed_attributes: total_attributes,
                indexed_files: PREVIEW_SLICE,
                total_attributes: Some(total_attributes),
                total_files: Some(total_files),
                complete: false,
            },
            reset: true,
            cached_data: None,
        })
        .unwrap();

        app.pump_index_updates();
        app.pump_search_results();
        assert!(
            app.filtered_len() > 0,
            "expected preview results to be visible during indexing"
        );

        app.throbber_state.calc_next();

        let mut terminal = Terminal::new(TestBackend::new(100, 30)).unwrap();
        terminal.draw(|frame| app.draw(frame)).unwrap();

        let view = {
            let backend = terminal.backend();
            backend.to_string()
        };

        insta::assert_snapshot!(
            "massive_filesystem_initial_load_shows_preview_snapshot",
            view
        );
    }
}
