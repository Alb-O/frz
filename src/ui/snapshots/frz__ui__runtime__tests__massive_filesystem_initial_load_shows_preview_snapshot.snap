---
source: src/ui/runtime.rs
assertion_line: 287
expression: view
---
" File search >                 ⠯ Indexed attributes: 246 • Indexed Files: 512/125000  Tags   Files  "
"   Path                    Tags           Score   Unable to preview dir_0000/file_000000.txt:       "
"  ─────────────────────────────────────────────── 'dir_0000/file_000000.txt': No such file or       "
" ▶ …r_0000/file_000000.txt *.txt, dir_00… 0       directory (os error 2)                            "
"   …r_0000/file_000001.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000002.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000003.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000004.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000005.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000006.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000007.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000008.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000009.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000010.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000011.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000012.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000013.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000014.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000015.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000016.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000017.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000018.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000019.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000020.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000021.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000022.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000023.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000024.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000025.txt *.txt, dir_00… 0                                                         "
"   …r_0000/file_000026.txt *.txt, dir_00… 0                                                         "
